src/python/pants/backend/docker/target_types.py

```diff
@@ -0,0 +1,42 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from pants.engine.target import COMMON_TARGET_FIELDS, Dependencies, Sources, StringField, Target
+
+
+class DockerImageSources(Sources):
+    default = ("Dockerfile",)
+    expected_num_files = 1
+    help = "Name of the Dockerfile to use when building the Docker image."
+
+
+class DockerImageVersion(StringField):
+    alias = "version"
+    default = "latest"
+    help = "Image tag to apply to built images."
+
+
+class DockerDependencies(Dependencies):
+    pass
+
+
+class DockerBuildRoot(StringField):
+    alias = "build_root"
+    default = "."
+    help = (
+        "Root directory for Docker build context. Default is to use the directory of the `BUILD` "
```

Comment:
So is the right way to think about it: we create a chroot containing just your transitive dependencies, and then a subset of that is your docker context?

Reply:
Yes, I've been working this over in my head. Not saying I got it right yet.

Also, any pex binaries you depend on, will be copied into your context, regardless.. which feels.. kind of odd. (in the current implementation around the corner)

What we have is, the chroot with the sources (i.e. the `Dockerfile` and any referenced `files` and `resources`) and all those that live in that subtree from the `BUILD` file, will be part of the docker context.

What I struggle with, is how to handle when you add dependencies to resources outside of your subtree.. as those will end up in the chroot, but will not be part of the docker context, so you can't use them.

So, if we put the docker context at the root of the chroot, then you can get at all files, but then you'll have to add the path to your Dockerfile when referencing your files or resources, which feels unintuitive.

A better approach could be to simply drop the prefix to the Dockerfile from the chroot, for all files in there (digest `RemovePrefix`) hoping to not have name conflicts...

So, best case, is that we don't need the build_root field at all.